cf/weights.py

```diff
@@ -0,0 +1,1903 @@
+                `True` if weights were created, otherwise `False`.
+
+        """
+        # ------------------------------------------------------------
+        # Data weights
+        # ------------------------------------------------------------
+        field_data_axes = f.get_data_axes()
+
+        if axes is not None:
+            if isinstance(axes, (str, int)):
+                axes = (axes,)
+
+            if len(axes) != w.ndim:
+                raise ValueError(
+                    "'axes' parameter must provide an axis identifier "
+                    f"for each weights data dimension. Got {axes!r} for "
+                    f"{w.ndim} dimension(s)."
+                )
+
+            iaxes = [
+                field_data_axes.index(f.domain_axis(axis, key=True))
+                for axis in axes
+            ]
+
+            if data:
+                for i in range(f.ndim):
+                    if i not in iaxes:
+                        w = w.insert_dimension(position=i)
+                        iaxes.insert(i, i)
+
+                w = w.transpose(iaxes)
+
+                if w.ndim > 0:
+                    while w.shape[0] == 1:
+                        w = w.squeeze(0)
+
+        else:
+            iaxes = list(range(f.ndim - w.ndim, f.ndim))
+
+        if not (components or methods):
+            if not f._is_broadcastable(w.shape):
+                raise ValueError(
+                    f"The 'Data' weights (shape {w.shape}) are not "
+                    "broadcastable to the field construct's data "
+                    f"(shape {f.shape})."
+                )
+
+            axes0 = field_data_axes[f.ndim - w.ndim :]
+        else:
+            axes0 = [field_data_axes[i] for i in iaxes]
+
+        for axis0 in axes0:
+            if axis0 in weights_axes:
+                raise ValueError(
+                    "Multiple weights specified for "
+                    f"{f.constructs.domain_axis_identity(axis0)!r} axis"
+                )
+
+        if methods:
+            weights[tuple(axes0)] = "custom data"
+        else:
+            weights[tuple(axes0)] = w
+
+        weights_axes.update(axes0)
+        return True
+
+    @classmethod
+    def field(cls, f, field_weights, weights, weights_axes):
+        """Create weights from other field constructs.
+
+        .. versionadded:: 3.16.0
+
+        :Parameters:
+
+            f: `Field`
+                The field for which the weights are being created.
+
+            field_weights: sequence of `Field`
+                The field constructs from which to create weights.
+
+            {{weights weights: `dict`}}
+
+            {{weights weights_axes: `set`}}
+
+        :Returns:
+
+            `bool`
+                `True` if weights were created, otherwise `False`.
+
+        """
+        s = f.analyse_items()
+
+        domain_axes = f.domain_axes(todict=True)
+        for w in field_weights:
+            t = w.analyse_items()
+            # TODO CHECK this with org
```

Comment:
Does something need doing here (for this PR), RE the TODO?

Reply:
I think that that comment is so many years old we can just remove it :)